RevisionDiff: Stage/Unstage selected lines

gerhardol opened the pull request:
Fixes #7338 

Based on #8708 

## Proposed changes

Move handling of stage/unstage/reset line patching from FormCommit to FileViewer.
Instead of FormCommit adding menu items to FileViewer, handle
the actions directly in FileViewer, including the hotkey handling.
Also hotkey for general stage/reset lines (previously cherry-pick and revet line patches).
This also aligns the stage/reset menu items to other forms.

Apply the line patching also to RevDiff, supporting line patches for
worktree/index properly, including hotkey handling.
When manipulating a diff involving an artificial commit, refresh the diff

Some code cleanup, like using ViewChangesAsync consistently for diffs/patches

This is a step to avoid using FormCommit to prepare commits

## Screenshots

For WorkTree, Similar for Index
Normal commits gets hotkeys and changed Copy grouping too

### Before

![image](https://user-images.githubusercontent.com/6248932/75594925-cd9fa580-5a8a-11ea-88aa-b3b370febf8b.png)

![image](https://user-images.githubusercontent.com/6248932/75594888-acd75000-5a8a-11ea-9f9d-6d4f51fcf596.png)


### After

![image](https://user-images.githubusercontent.com/6248932/103446613-e3b95c00-4c81-11eb-96ed-e07fd9ae7194.png)

![image](https://user-images.githubusercontent.com/6248932/103446621-0481b180-4c82-11eb-9626-645cabd46dc3.png)


## Test methodology <!-- How did you ensure quality? -->

Manual

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

gerhardol commented:
CodeFactor complains about complexity for a function with case (3 more added by new hotkeys). I do not see a good way to rewrite that.

RussKie commented:
Don't worry about it

On Mon, Mar 2, 2020, 10:57 AM Gerhard Olsson <notifications@github.com>
wrote:

> CodeFactor complains about complexity for a function with case (3 more
> added by new hotkeys). I do not see a good way to rewrite that.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/7825?email_source=notifications&email_token=ABBTEXS3CMACUUG5ZGTY7VTRFLY65A5CNFSM4K6CQWJ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENNQGTI#issuecomment-593167181>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXVL7YPCWH2GYWBEZETRFLY65ANCNFSM4K6CQWJQ>
> .
>

gerhardol commented:
Updated for #7831 
(to be squashed on next update)
I considered a nicer solution than enabling the stage/unstage menu item in TextLoaded, but will wait for that for  after #4154 (that partly depends on this PR).
Build fails on translation, I expect this  to go in after #7877.

@mstv @RussKie Can you please review?

mstv commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -1063,7 +1070,8 @@ private void DiffText_TextLoaded(object sender, EventArgs e)
         {
-            if (DiffFiles.SelectedItem == null ||
+            var item = DiffFiles.SelectedItem;
+            if (item == null ||
```
Nit: This modification is unnecessary.

gerhardol replied:
It is an effect of my backported changes. I can revert it...

mstv replied:
Keep `item` if you are going to use it in a follow up PR.

gerhardol replied:
I reverted that, there will be other changes to the assignment

gerhardol commented:
> Stage/untage file/selected lines seem to work well. But the hotkeys are not shown for the FileStatusList.

I cannot why they do not appear...

> The next thing I really like to have, is that the hotkeys work from the RevisionGrid, too, because it gets focused when switching the artificial commits.

and/or hotkeys for switching between worktree/index, or global keys to select worktree/index and view RevDiff
But that is later, I would like to finish other open items first...

mstv commented on GitUI/CommandsDialogs/RevisionDiffControl.Designer.cs:
```diff
@@ -245,6 +245,7 @@ private void InitializeComponent()
             this.stageFileToolStripMenuItem.Name = "stageFileToolStripMenuItem";
             this.stageFileToolStripMenuItem.Size = new System.Drawing.Size(296, 22);
             this.stageFileToolStripMenuItem.Text = "Stage file(s)";
+            this.stageFileToolStripMenuItem.ShortcutKeyDisplayString = GetShortcutKeyDisplayString(Command.StageSelectedFile);
```
Add the following to `ReloadHotkeys` instead:
```C#
            resetFileToParentToolStripMenuItem.ShortcutKeyDisplayString = GetShortcutKeyDisplayString(Command.ResetSelectedFiles);
            stageFileToolStripMenuItem.ShortcutKeyDisplayString = GetShortcutKeyDisplayString(Command.StageSelectedFile);
            unstageFileToolStripMenuItem.ShortcutKeyDisplayString = GetShortcutKeyDisplayString(Command.UnStageSelectedFile);
```

gerhardol replied:
This also fixes that the shortcuts were not shown and makes it reasonable to allow reset file by shortcut (still no popup)
![image](https://user-images.githubusercontent.com/6248932/77250315-73d46a80-6c47-11ea-8f0e-9613dd521a63.png)

mstv commented:
> global keys to select worktree/index and view RevDiff

👍 That's the use case. Of course, for later.

mstv commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -998,14 +1001,25 @@ private bool UnStageSelectedFile()
+
+                // Reset to first (parent)
+                ResetSelectedItemsTo(false);
+                return true;
```
readability can be improved
```suggestion
                if (parents.Count == 1 && CanResetToRevision(parents.FirstOrDefault()))
                {
                    // Reset to first (parent)
                    ResetSelectedItemsTo(actsAsChild: false);
                    return true;
                }
```

gerhardol replied:
I can agree on the readability and it follows the pattern better, but @RussKie usually requests as early exits as possible...

gerhardol replied:
will change, calculate revisions before first if to reduce depth (readability is more important than a few CPU cycles saved when line patching)

gerhardol commented on GitUI/Editor/FileViewer.cs:
```diff
@@ -336,6 +338,72 @@ public ToolStripMenuItem AddContextMenuEntry(string text, EventHandler toolStrip
+        /// <summary>
+        /// If the file viewer menu supports line patch editing
+        /// </summary>
+        public bool SupportLinePatching { get; set; }
```
ok

gerhardol commented on GitUI/Editor/FileViewer.cs:
```diff
@@ -336,6 +338,72 @@ public ToolStripMenuItem AddContextMenuEntry(string text, EventHandler toolStrip
+
+            revertSelectedLinesToolStripMenuItem.Click -=
+                revertSelectedLinesToolStripMenuItem_Click;
+            revertSelectedLinesToolStripMenuItem.Click +=
```
I have gotten used to code in Groovy, where it is preferred to have the operator at the end of the line. Will change

gerhardol commented on GitUI/Editor/FileViewer.cs:
```diff
@@ -336,6 +338,72 @@ public ToolStripMenuItem AddContextMenuEntry(string text, EventHandler toolStrip
+        public void CherryPickContextMenuEntry_Visible()
+        {
+            cherrypickSelectedLinesToolStripMenuItem.Visible =
+                SupportLinePatching = true;
```
Separating assign to SupportLinePatching, can remove multiassign. Rename

gerhardol commented on GitUI/Editor/FileViewer.cs:
```diff
@@ -360,16 +428,11 @@ private void SetVisibilityDiffContextMenu(bool visibleTextFile, [CanBeNull] stri
             cherrypickSelectedLinesToolStripMenuItem.Visible =
                 revertSelectedLinesToolStripMenuItem.Visible =
+                    SupportLinePatching =
                     visibleTextFile && fileExists && !Module.IsBareRepository();
```
see above

gerhardol commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -928,5 +943,366 @@ public void SwitchFocus(bool alreadyContainedFocus)
+        {
+            if (!string.IsNullOrEmpty(output))
+            {
+                MessageBox.Show(this, output + "\n\n" + DiffText.Encoding.GetString(patch), "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
```
I have noticed that too but it should be separate - in some other places too
(Good to comment unrelated while reviewing, just to be clear what should be handled here)

gerhardol replied:
#7893 created, need to rebase this PR and change this before merge

gerhardol commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -928,5 +943,366 @@ public void SwitchFocus(bool alreadyContainedFocus)
+            else if (DiffFiles.Focused && DiffFiles.Revision.ObjectId == ObjectId.WorkTreeId)
+            {
+                // Only stage files for WorkTree (even if lines can be cherry-picked/staged)
+                StageFileToolStripMenuItemClick(this, null);
```
Changing to the common pattern to call shared code instead

gerhardol commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -928,5 +943,366 @@ public void SwitchFocus(bool alreadyContainedFocus)
+
+            if (DiffText.ContainsFocus && DiffText.SupportLinePatching)
+            {
+                StageSelectedLinesToolStripMenuItemClick(this, null);
```
Changing to the common pattern to call shared code instead

gerhardol commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -928,5 +943,366 @@ public void SwitchFocus(bool alreadyContainedFocus)
+            }
+            else if (DiffFiles.Focused)
+            {
+                UnstageFileToolStripMenuItemClick(this, null);
```
Changing to the common pattern to call shared code instead

gerhardol commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -928,5 +943,366 @@ public void SwitchFocus(bool alreadyContainedFocus)
+                return false;
+            }
+
+            return true;
```
ok

mstv commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -928,5 +953,357 @@ public void SwitchFocus(bool alreadyContainedFocus)
+            else if (DiffFiles.Focused && parents.Count == 1 && CanResetToRevision(parents.FirstOrDefault()))
+            {
+                // Reset to first (parent)
+                ResetSelectedItemsTo(false);
```
```suggestion
                ResetSelectedItemsTo(actsAsChild: false);
```

gerhardol replied:
ok

gerhardol commented:
> R works for file as expected now. I guess I had not built everything last time.

I have seen this too, since around the time SDK style project was merged. (I did report this in a comment, probably my worst issue description ever). It seem to occur more often if I have not saved before Start/F5, but there has been situations where I thought I saved too.

Dont know how to proceed with that.

gerhardol commented:
There are some things I am not so happy about with this PR.
Some changes will be simpler with changes that depends on the following:
https://github.com/gitextensions/gitextensions/issues/4154#issuecomment-602254811

Replace ViewCurrentChanges with ViewChangesAsync - to unify

Another is the handling of SelectedDiff.CherryPickContextMenuEntry_Visible(); in TextLoaded. If the file to view is stored in FileViewer as GitItemStatusWithParent then that handling can be built in, as well as the patch handling that currently is slightly duplicated in FileViewer, FormCommit and RevisionDiffControl.

gerhardol commented:
Squashed review comments and changes due to #7831, rebased on #7868 (to be merged tomorrow, will conflict). If no further review I plan to initiate a merge in two days.

gerhardol commented:
> Go ahead!

I will give our mighty leader @RussKie a chance to review too, even if he cannot do everything.
(will mark this as off topic myself...)

RussKie commented:
I'll do it in a day or two, if that's ok

On Mon, 23 Mar 2020 at 09:14, Gerhard Olsson <notifications@github.com>
wrote:

> Go ahead!
>
> I will give our mighty leader @RussKie <https://github.com/RussKie> a
> chance to review too, even if he cannot do everything.
> (will mark this as off topic myself...)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/7825#issuecomment-602284869>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXRGVRYXYBXCI4KYASDRI2EWDANCNFSM4K6CQWJQ>
> .
>

mstv commented:
> `R` works for file as expected now. I guess I had not built everything last time.

This is no build issue. Sometimes, `R` works, sometimes not. Then a sound is played as if it was an invalid or not supported key. The menu item `Reset to | First` has always worked. (I have no clue yet what might be the difference in the situations. Explicitly focusing by clicking the FileStatusList does not help.)

gerhardol commented:
> This is no build issue. Sometimes, R works, sometimes not. Then a sound is played as if it was an invalid or not supported key. The menu item Reset to | First has always worked. (I have no clue yet what might be the difference in the situations. Explicitly focusing by clicking the FileStatusList does not help.)

If no other solution, I suggest disabling the shortcut (as it was initially).

mstv commented:
We'll make it work. I use it often enough and will watch it. I don't like fiddling with the submenu.

mstv commented:
`R` works for real commits and for `Commit Index` but not for `Working Directory`.

gerhardol commented:
> R works for real commits and for Commit Index but not for Working Directory.

For changes to tracked files, it works for me. The command that is executed is the same as for the menu item (just that the menu item is disabled when you cannot reset, like to worktree).

For new files this works as for the reset menu: It does not
This requires some special handling. I would prefer to handle that in #4154 where the first/selected commits is known in the selected item.
A PR for #4154 depends on this PR, that awaits a review from @RussKie 

Adding a commit to simplify merge with #7887 (will be merged first)

gerhardol commented:
> I'm not content with the new API. I'll need to think through it

I plan further cleanups of the API after #7899, but I guess you do not want any bigger PRs than this?

RussKie commented:
> * Refactor the FormCommit handling from adding menu items to FileViewer,
>   to reusing cherry-pick/revert items, so the FileViewer menu is aligned
>   The same setup used for the RevisionDiff menus

I'm not happy with this, and I'm not liking the proposed API.
By reusing we're not saving much, but increasing a confusion on multiple levels.
Is there any reason why we can't add new menu items, that have set captions/tooltips/shortcuts/etc, and make them visible/hidden as required?

gerhardol commented:
> Is there any reason why we can't add new menu items, that have set captions/tooltips/shortcuts/etc, and make them visible/hidden as required?

One reason is consistency, the menu items show up in different positions in different forms.

Then the logic for enabling disabling need to be updated too, FormCommit/RevDiff will have to implement the refresh privately.
(In the current implementation, FormCommit has separate interface to FileViewer, that is to removed in follow up PR with  0a7543701d0119201412b4b9cc075bd084dbc916)

After #7899 I plan to unify the line patch handling and move the logic to FileViewer
(FileViewer will have to save a GitItemStatusWithParents if available and implement most of the logic that are parallel implemented in FileViewer, FormCommit, RevisionDiffControl now)

RussKie commented:
Ok, if there is a plan to do it in a near future, I'm not going to block it. Though please make these at least "internal".

gerhardol commented:
> Ok, if there is a plan to do it in a near future, I'm not going to block it. Though please make these at least "internal".

Can you outline what you want to see in a final API?

My plan is to keep the separate event handlers, but have most code common in FileViewer. RevDiff and FormCommit still some handling to refresh.

RussKie commented:
Can we have an overarching discussion about what we want to achieve in the
target state?
What API we have, what's missing and what needs changing.

There's a lot of change and churn atm, and I'm struggling to keep track of
it.
I don't want to have new methods added for the sake of adding. I'm sure we
can clean the lot as well.

On Mon, Mar 30, 2020, 8:00 AM Gerhard Olsson <notifications@github.com>
wrote:

> Ok, if there is a plan to do it in a near future, I'm not going to block
> it. Though please make these at least "internal".
>
> Can you outline what you want to see in a final API?
>
> My plan is to keep the separate event handlers, but have most code common
> in FileViewer. RevDiff and FormCommit still some handling to refresh.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/7825#issuecomment-605700668>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXRIAFM7TJS7GGOBM73RJ6ZGPANCNFSM4K6CQWJQ>
> .
>

gerhardol commented:
After #7899
FileViewer.ViewChanges(GitStatusItemWithParents item)
FileViewer can use the item to determine what line stage options that are available as well as the names and icons.
For the EventHandlers, forms may need to perform actions (at least FormCommit and RevDiff to refresh the artificial commits) so the implementation of the event viewer still need to be local for those with specific requirements but most of the code to create and apply patches can be common.

mstv commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -902,5 +927,357 @@ public void SwitchFocus(bool alreadyContainedFocus)
+                StageOrCherryPickSelectedLines();
+                return true;
+            }
+            else if (DiffFiles.Focused && DiffFiles.Revision.ObjectId == ObjectId.WorkTreeId)
```
nit: `else` can be omitted
```suggestion
            
            if (DiffFiles.Focused && DiffFiles.Revision.ObjectId == ObjectId.WorkTreeId)
```

gerhardol replied:
ok for all

mstv commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -902,5 +927,357 @@ public void SwitchFocus(bool alreadyContainedFocus)
+                StageOrCherryPickSelectedLines();
+                return true;
+            }
+            else if (DiffFiles.Focused)
```
ditto

mstv commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -902,5 +927,357 @@ public void SwitchFocus(bool alreadyContainedFocus)
+                ResetOrRevertSelectedLines();
+                return true;
+            }
+            else if (DiffFiles.Focused && parents.Count == 1 && CanResetToRevision(parents.FirstOrDefault()))
```
ditto

mstv commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -902,5 +927,357 @@ public void SwitchFocus(bool alreadyContainedFocus)
+            if (item.IsNew)
+            {
+                var treeGuid = currentItemStaged ? item.TreeGuid?.ToString() : null;
+                patch = PatchManager.GetSelectedLinesAsNewPatch(Module, item.Name,
+                    DiffText.GetText(), DiffText.GetSelectionPosition(),
+                    DiffText.GetSelectionLength(), DiffText.Encoding, false, DiffText.FilePreamble, treeGuid);
+            }
+            else
+            {
+                patch = PatchManager.GetSelectedLinesAsPatch(
+                    DiffText.GetText(),
+                    DiffText.GetSelectionPosition(), DiffText.GetSelectionLength(),
+                    currentItemStaged, DiffText.Encoding, item.IsNew);
```
Our new style is to have all arguments on one line or on single lines.

mstv commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -902,5 +927,357 @@ public void SwitchFocus(bool alreadyContainedFocus)
+            byte[] patch;
+            if (reverse)
+            {
+                patch = PatchManager.GetResetWorkTreeLinesAsPatch(
+                    Module, DiffText.GetText(),
+                    DiffText.GetSelectionPosition(), DiffText.GetSelectionLength(), DiffText.Encoding);
+            }
+            else
+            {
+                patch = PatchManager.GetSelectedLinesAsPatch(
+                    DiffText.GetText(),
+                    DiffText.GetSelectionPosition(), DiffText.GetSelectionLength(),
+                    false, DiffText.Encoding, false);
```
ditto arguments

mstv commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -902,5 +927,357 @@ public void SwitchFocus(bool alreadyContainedFocus)
+            if (item.IsNew)
+            {
+                var treeGuid = currentItemStaged ? item.TreeGuid?.ToString() : null;
+                patch = PatchManager.GetSelectedLinesAsNewPatch(Module, item.Name,
+                    DiffText.GetText(), DiffText.GetSelectionPosition(), DiffText.GetSelectionLength(),
+                    DiffText.Encoding, true, DiffText.FilePreamble, treeGuid);
+            }
+            else if (currentItemStaged)
+            {
+                patch = PatchManager.GetSelectedLinesAsPatch(
+                    DiffText.GetText(),
+                    DiffText.GetSelectionPosition(), DiffText.GetSelectionLength(),
+                    currentItemStaged, DiffText.Encoding, item.IsNew);
+            }
+            else
+            {
+                patch = PatchManager.GetResetWorkTreeLinesAsPatch(Module, DiffText.GetText(),
+                    DiffText.GetSelectionPosition(), DiffText.GetSelectionLength(), DiffText.Encoding);
```
ditto arguments

gerhardol commented:
plan to base this on top of #7899 (when that has been reviewed) and change the API to get @RussKie  happier

gerhardol commented:
Added a cleanup that should be done since merging #7926, will continue to do also if not it will be merged soon

mstv commented on GitUI/CommandsDialogs/FormCommit.cs:
```diff
@@ -1828,16 +1819,27 @@ private void Unstage(bool canUseUnstageAll = true)
             }
 
-            return null;
+            var workTreeRev = new GitRevision(ObjectId.WorkTreeId)
+            {
+                ParentIds = new[] { ObjectId.IndexId }
+            };
```
Put this on one line as above for `indexRev`, please.

gerhardol replied:
squashed

gerhardol commented:
Rebased on master, with review comments and adaption for #7926 

This branch based on #7899: https://github.com/gerhardol/gitextensions/tree/feature/i7338-revdiff-stage-2
Will await a review by RussKie before preparing updates here, there will be too many changes otherwise

RussKie commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -902,5 +927,383 @@ public void SwitchFocus(bool alreadyContainedFocus)
+        /// Similar to FileViewer.applySelectedLines()
+        /// </summary>
+        /// <param name="reverse">set if patches is to be reversed</param>
+        private void applySelectedLines(bool reverse)
```
```suggestion
        private void ApplySelectedLines(bool reverse)
```

gerhardol replied:
OK for all

RussKie commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -902,5 +927,383 @@ public void SwitchFocus(bool alreadyContainedFocus)
+            applySelectedLines(false);
+        }
+
+        /// <summary>
+        /// apply patches
+        /// Similar to FileViewer.applySelectedLines()
+        /// </summary>
+        /// <param name="reverse">set if patches is to be reversed</param>
```
```suggestion
        /// <summary>
        /// Apply patches.
        /// Similar to FileViewer.ApplySelectedLines()
        /// </summary>
        /// <param name="reverse"><see langword="true"/> if patches is to be reversed; otherwise <see langword="false"/>.</param>
```

RussKie commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -902,5 +927,383 @@ public void SwitchFocus(bool alreadyContainedFocus)
+        /// Hotkey handler
+        /// </summary>
+        /// <returns>true if executed</returns>
+        private bool UnStageSelectedFile()
```
```suggestion
        private bool UnstageSelectedFile()
```

RussKie commented on GitUI/Editor/FileViewer.cs:
```diff
@@ -383,38 +385,98 @@ public ToolStripMenuItem AddContextMenuEntry(string text, EventHandler toolStrip
+        /// </summary>
+        /// <param name="cherryToolStripItemClick">Cherry pick event handler</param>
+        /// <param name="resetToolStripItemClick">Reset event handler</param>
+        public void CherryPickContextMenuEntry_OverrideClick(EventHandler cherryToolStripItemClick, EventHandler resetToolStripItemClick)
```
I've been staring at the override API, but I am really struggling to grasp it.

Do you think we could explore a different way to express the desired functionality?
E.g. something like this:

```cs
	// FileViewer.cs

        public class OverrideEvents
        {
            public EventHandler CherryPickSelected;
            public EventHandler ResetSelected;
        }

        public OverrideEvents Overrides { get; } = new OverrideEvents();


        private void SetVisibilityDiffContextMenu(bool isDiff, [CanBeNull] string fileName)
        {
			// ...

            cherrypickSelectedLinesToolStripMenuItem.Visible = SupportLinePatching && Overrides.CherryPickSelected != null;
            revertSelectedLinesToolStripMenuItem.Visible = SupportLinePatching && Overrides.ResetSelected != null;
        }
```

```cs
	// FormCommit.cs

            SelectedDiff.Overrides.CherryPickSelected += StageSelectedLinesToolStripMenuItemClick;
            SelectedDiff.Overrides.ResetSelected += ResetSelectedLinesToolStripMenuItemClick;
```

RussKie commented:
I'm not sure we're approaching the implementation from the right angle. `FileViewer`'s encapsulation feels very weak, we appear to be manipulating the same menu items over and over by flipping text and icons in the consuming controls... 
This makes it very hard to reason about what and when sets what and track where it gets handled.

Have you considered having "Cherry pick", "Stage selected", "Unstage selected", "Reset selecte", etc. menus, with the same appearance and shortcuts, and have `FileViewer` expose simple API to show/hide these menus, e.g.: 
```cs
bool ShowCherryPickSelected { get; set }
bool ShowStageSelected { get; set }
bool ShowUnstageSelected { get; set }
bool ShowResetSelected { get; set }
```

Alternative handles can be wired up via something like the `Overrides` I proposed in https://github.com/gitextensions/gitextensions/pull/7825/commits/f36a98bb0ef6a305f804c379145e4057c33477be.

This way we will have a bigger menu and more code in `FileViewer` (-) but it will be constrained to `FileViewer` alone (+), and all its consumers will interact with the menu via a simplified API surface (+).

gerhardol commented:
> . FileViewer's encapsulation feels very weak, we appear to be manipulating the same menu items over and over by flipping text and icons in the consuming controls...

FileViewer has not much details of what it is to show right now. Basically the Forms ask to view Text or Changes, but FileViewer has no information of what revisions to display (except in current master where FormCommit has special handling, a much bigger hack than the API in this PR). Menu Items are somehow adjusted to what is displayed (again, FormCommit has the special hooks to adjust the presentation). RevDiff/Commit uses the Event Text_Changed to enable the correct menu items.

After the text is displayed, FileViewer has no knowledge of what it has displayed. The information about revisions and the file to present is required to perform the actions like staging lines. However, the forms still need to do private actions, like refreshing the views, done from the menu event handlers.

What you are proposing in https://github.com/gitextensions/gitextensions/pull/7825#issuecomment-616136617 is basically to not adjust the CherryPick/Revert menu items but to add extra stage/unstage/reset items. SetVisibilityDiffContextMenu would reset Enable to default but RevDiff/Commit Text_Loaded will then adjust the enabled menus. The code base will increase some though.

Similarly, f36a98bb0ef6a305f804c379145e4057c33477be is handling the event handlers differently. (But event handlers need to be added for all five menu items).

These changes will not change the responsibilities of the modules, just a patchup of the API.
Regardless of these changes, it may be possible to share a little more of the patch and git-apply handling, currently slightly separate in FileViewer/RevDiff/Commit

But before doing these changes, I would like to agree on if the structure should be changed or not.
I will not change many times again, like for other changes and commented on here.

If FileViewer knows what to display, like ```ShowPatch(GitItemStatusWithParents item)``` with first/second revision according to #7899 (some Forms will still use ShowChanges(String text) ). Then FileViewer can enable the menu items and implement the git-apply handling. The event handlers for the staging actions are still needed (possibly to refresh the forms).
But then this PR is dependent of #7899 and at least part of that PR should be merged first.

gerhardol commented:
Squashed on top of master (now that #7899 is merged)
Set to WIP as this will not be merged before 3.4 and some changes will be done.

mstv commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -948,5 +973,372 @@ public void SwitchFocus(bool alreadyContainedFocus)
+
+            return (item.SecondRevision.ObjectId == ObjectId.WorkTreeId
+                    || item.SecondRevision.ObjectId == ObjectId.IndexId)
+                   && item.SecondRevision.FirstParentGuid == item.FirstRevision?.ObjectId;
```
Do you still consider this WIP -- except this minor merge conflict?
```suggestion
                   && item.SecondRevision.FirstParentId == item.FirstRevision?.ObjectId;
```

gerhardol replied:
This is high priority for me for next release. I get very frustrated when I am not using this and have to use FormCommit...

RussKie requested changes, he was not satisfied with keeping the responsibilities as they are before this PR. 
I therefore have have to rewrite some, which I have not started with yet. (Wanted the changes leading up to 3.4 to settle and not spend all time rebasing, then I want to finish other tasks that are closer to completion.)

gerhardol replied:
I had already changed this in my wip branch...

gerhardol commented:
Just rebased on master (this is really what I am using myself).
Still trying to rewrite this, going slow. Quite some new code needed (probably more than saved by using common snippets).

mstv commented:
👍 works well again (shortcuts), (waiting for the refactoring).

gerhardol commented:
> 👍 works well again (shortcuts), (waiting for the refactoring).

The merge problems was in the private branch only that I had not pushed in a while.
I am not so inspired to rewrite this, it takes some time...

gerhardol commented:
Rewritten the existing functionality (and thereby this PR) after @RussKie requests.
Still WIP as  #8708 is not merged yet - I actually would like a comment on this PR before merging that PR.
The code is ready for a review though.

mstv reviewed (commented):
LGTM. Only nits. Will run it later.

gerhardol commented on GitUI/CommandsDialogs/RevisionDiffControl.cs:
```diff
@@ -78,17 +79,17 @@ public void UICommands_PostRepositoryChanged(object sender, GitUIEventArgs e)
         }
 
-        public void RefreshArtificial()
+        public bool RefreshArtificial()
```
I was using it in an intermediate version, to check if ArtificialCommits were affected, but changed to only run the callback for artificial. There were some thoughts if this wasenough though and I kept this.
So I will revert this change.

gerhardol commented on GitUI/Editor/FileViewer.cs:
```diff
@@ -1665,7 +1877,10 @@ internal enum Commands
             NextChange = 6,
             PreviousChange = 7,
             NextOccurrence = 10,
-            PreviousOccurrence = 11
+            PreviousOccurrence = 11,
+            StageLines = 12,
+            UnstageLines = 13,
+            ResetLines = 14
```
OK

gerhardol commented on GitUI/CommandsDialogs/FormCommit.cs:
```diff
@@ -751,9 +723,9 @@ private bool UnStageSelectedFile()
+            else if (SelectedDiff.ContainsFocus && _currentItemStaged)
             {
-                StageSelectedLinesToolStripMenuItemClick(this, null);
+                SelectedDiff.StageSelectedLines(false);
```
I prefer to not have too many interface functions, so I go with the first suggestion.

gerhardol commented:
Rebased on latst #8708 , review comments addressed and a fixup of my own

gerhardol commented:
Squashed review comments (but not my unreviewed changes) and rebased on master, removed WIP now when #8708 is merged

RussKie commented:
Since @mstv is happy, let's merge it :)

gerhardol commented:
Will merge after #8734 

Now I only miss a way to change the commit message for Index commit and a commit button to skip the modal commit dialog in most situations

gerhardol commented:
Squashed and rebased after #8734 with a minor removal ofunused code in FormCommit:ShowChanges()

gerhardol commented:
@msftbot merge in 3 days
